sdks/go/pkg/beam/core/runtime/harness/monitoring.go

```diff
@@ -0,0 +1,195 @@
+	transforms := make(map[string]*fnpb.Metrics_PTransform)
+	metrics.Extractor{
+		SumInt64: func(l metrics.Labels, v int64) {
+			pb := getTransform(transforms, l)
+			pb.User = append(pb.User, &fnpb.Metrics_User{
+				MetricName: toName(l),
+				Data: &fnpb.Metrics_User_CounterData_{
+					CounterData: &fnpb.Metrics_User_CounterData{
+						Value: v,
+					},
+				},
+			})
+		},
+		DistributionInt64: func(l metrics.Labels, count, sum, min, max int64) {
+			pb := getTransform(transforms, l)
+			pb.User = append(pb.User, &fnpb.Metrics_User{
+				MetricName: toName(l),
+				Data: &fnpb.Metrics_User_DistributionData_{
+					DistributionData: &fnpb.Metrics_User_DistributionData{
+						Count: count,
+						Sum:   sum,
+						Min:   min,
+						Max:   max,
+					},
+				},
+			})
+		},
+		GaugeInt64: func(l metrics.Labels, v int64, t time.Time) {
+			ts, err := ptypes.TimestampProto(t)
+			if err != nil {
+				panic(err)
+			}
+			pb := getTransform(transforms, l)
+			pb.User = append(pb.User, &fnpb.Metrics_User{
+				MetricName: toName(l),
+				Data: &fnpb.Metrics_User_GaugeData_{
+					GaugeData: &fnpb.Metrics_User_GaugeData{
+						Value:     v,
+						Timestamp: ts,
+					},
+				},
+			})
+		},
+	}.ExtractFrom(p.Store)
+
+	// Get the MonitoringInfo versions.
+	var monitoringInfo []*ppb.MonitoringInfo
+	metrics.Extractor{
+		SumInt64: func(l metrics.Labels, v int64) {
+			monitoringInfo = append(monitoringInfo,
+				&ppb.MonitoringInfo{
+					Urn:    "beam:metric:user",
+					Type:   "beam:metrics:sum_int_64",
+					Labels: userLabels(l),
+					Data:   int64Counter(v),
+				})
+		},
+		DistributionInt64: func(l metrics.Labels, count, sum, min, max int64) {
+			monitoringInfo = append(monitoringInfo,
+				&ppb.MonitoringInfo{
+					Urn:    "beam:metric:user",
+					Type:   "beam:metrics:sum_int_64",
+					Labels: userLabels(l),
+					Data:   int64Distribution(count, sum, min, max),
+				})
+		},
+		GaugeInt64: func(l metrics.Labels, v int64, t time.Time) {
+			ts, err := ptypes.TimestampProto(t)
+			if err != nil {
+				panic(err)
+			}
+			monitoringInfo = append(monitoringInfo,
+				&ppb.MonitoringInfo{
+					Type:      "beam:metrics:latest_int_64",
+					Labels:    userLabels(l),
+					Data:      int64Counter(v),
+					Timestamp: ts,
+				})
+		},
+	}.ExtractFrom(p.Store)
+
+	// Get the execution monitoring information from the bundle plan.
+	if snapshot, ok := p.Progress(); ok {
+		// Legacy version.
+		transforms[snapshot.ID] = &fnpb.Metrics_PTransform{
+			ProcessedElements: &fnpb.Metrics_PTransform_ProcessedElements{
+				Measured: &fnpb.Metrics_PTransform_Measured{
+					OutputElementCounts: map[string]int64{
+						snapshot.Name: snapshot.Count,
+					},
+				},
+			},
+		}
+		// Monitoring info version.
+		monitoringInfo = append(monitoringInfo,
+			&ppb.MonitoringInfo{
+				Urn:  "beam:metric:element_count:v1",
+				Type: "beam:metrics:sum_int_64",
+				Labels: map[string]string{
+					"PCOLLECTION": snapshot.Name,
```

Comment:
This should be the PCollection id

Reply:
AFAICT There's no "input" to the Source transform, so can you confirm that it should be ID of the  Output PCollection?

Reply:
Chatted offline about this: There's only the output PCollection, and then this streamlines from "specific progress signal" to "ordinary PCollection metrics". I made TODOs for reminders when adding PCollection and DoFn metrics to the SDK.